File: be_camphub/src/main/java/org/camphub/be_camphub/mapper/ReturnRequestMapper.java
package org.camphub.be_camphub.mapper;

import org.camphub.be_camphub.dto.request.return_req.ReturnReqCreationRequest;
import org.camphub.be_camphub.dto.response.return_req.ReturnReqResponse;
import org.camphub.be_camphub.entity.ReturnRequest;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

@Mapper(componentModel = "spring", uses = MediaResourceMapper.class)
public interface ReturnRequestMapper {
    @Mapping(target = "lesseeName", ignore = true)
    @Mapping(target = "lessorName", ignore = true)
    @Mapping(target = "itemName", ignore = true)
    @Mapping(target = "evidenceUrls", source = "evidenceUrls", qualifiedByName = "toResponse")
    ReturnReqResponse entityToResponse(ReturnRequest returnRequest);

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "lesseeId", ignore = true) // set trong service
    @Mapping(target = "lessorId", ignore = true) // set trong service
    @Mapping(target = "status", constant = "PENDING")
    @Mapping(target = "adminNote", ignore = true)
    @Mapping(target = "resolvedAt", ignore = true)
    @Mapping(target = "createdAt", ignore = true) // Auto @CreatedDate
    @Mapping(target = "updatedAt", ignore = true) // Auto @LastModifiedDate
    @Mapping(target = "lessorConfirmedAt", ignore = true)
    @Mapping(target = "adminReviewedAt", ignore = true)
    @Mapping(target = "refundedAt", ignore = true)
    @Mapping(target = "evidenceUrls", source = "evidenceUrls", qualifiedByName = "fromRequest")
    ReturnRequest creationRequestToEntity(ReturnReqCreationRequest request);
}

File: fe_camphub/src/libs/core/mapping/return-request.mapper.ts
// return-request.mapper.ts
import { createMap, createMapper, forMember, mapFrom } from "@automapper/core";
import { pojos, PojosMetadataMap } from "@automapper/pojos";
import { ReturnRequest } from "../types";
import {
    ReturnReqCreationRequest,
    AdminDecisionRequest,
    LesseeSubmitReturnRequest,
    LessorConfirmReturnRequest
} from "../dto/request";
import { ReturnReqResponse } from "../dto/response";

// ------------------------
// 1. Tạo mapper
// ------------------------
export const returnRequestMapper = createMapper({
    strategyInitializer: pojos(),
});

// ------------------------
// 2. Metadata
// ------------------------
PojosMetadataMap.create<ReturnRequest>("ReturnRequest", {
    id: String,
    bookingId: String,
    lesseeId: String,
    lessorId: String,

    itemName: String,
    lesseeName: String,
    lessorName: String,

    reason: String,
    evidenceUrls: Array,
    status: String,
    note: String,
    adminNote: String,
    createdAt: String,
    resolvedAt: String,
    lessorConfirmAt: String,
    adminReviewedAt: String,
    refundedAt: String,
});
PojosMetadataMap.create<ReturnReqResponse>("ReturnReqResponse", {
    id: String,
    bookingId: String,
    lesseeId: String,
    lessorId: String,

    itemName: String,
    lesseeName: String,
    lessorName: String,

    reason: String,
    evidenceUrls: Array,
    status: String,
    note: String,
    adminNote: String,
    createdAt: String,
    resolvedAt: String,
    lessorConfirmAt: String,
    adminReviewedAt: String,
    refundedAt: String,
});
PojosMetadataMap.create<ReturnReqCreationRequest>("ReturnReqCreationRequest", {
    bookingId: String,
    reason: String,
    note: String,
    evidenceUrls: Array,
});
PojosMetadataMap.create<AdminDecisionRequest>("AdminDecisionRequest", {
    returnRequestId: String,
    isApproved: Boolean,
    adminNote: String,
});
PojosMetadataMap.create<LesseeSubmitReturnRequest>("LesseeSubmitReturnRequest", {
    returnRequestId: String,
    note: String,
    packingMediaUrls: Array,
});
PojosMetadataMap.create<LessorConfirmReturnRequest>("LessorConfirmReturnRequest", {
    bookingId: String,
});

// ------------------------
// 3. Mapping DTO response → Model
// ------------------------
createMap<ReturnReqResponse, ReturnRequest>(
    returnRequestMapper,
    "ReturnReqResponse",
    "ReturnRequest"
);

// ------------------------
// 4. Mapping Model → Creation Request DTO
// ------------------------
createMap<ReturnRequest, ReturnReqCreationRequest>(
    returnRequestMapper,
    "ReturnRequest",
    "ReturnReqCreationRequest",
);

// ------------------------
// 5. Mapping Model → Patch / Update Requests (Admin, Lessee, Lessor)
// ------------------------
createMap<ReturnRequest, AdminDecisionRequest>(
    returnRequestMapper,
    "ReturnRequest",
    "AdminDecisionRequest",
    // map adminNote mặc định nếu null
);

createMap<ReturnRequest, LesseeSubmitReturnRequest>(
    returnRequestMapper,
    "ReturnRequest",
    "LesseeSubmitReturnRequest",
    forMember(d => d.packingMediaUrls, mapFrom(s => s.evidenceUrls || []))
);

createMap<ReturnRequest, LessorConfirmReturnRequest>(
    returnRequestMapper,
    "ReturnRequest",
    "LessorConfirmReturnRequest",
    forMember(d => d.bookingId, mapFrom(s => s.bookingId))
);

// ------------------------
// 6. Export API
// ------------------------
export const returnRequestMap = {
    fromResponse: (response: ReturnReqResponse): ReturnRequest =>
        returnRequestMapper.map<ReturnReqResponse, ReturnRequest>(
            response,
            "ReturnReqResponse",
            "ReturnRequest"
        ),
    toCreationRequest: (model: ReturnRequest): ReturnReqCreationRequest =>
        returnRequestMapper.map<ReturnRequest, ReturnReqCreationRequest>(
            model,
            "ReturnRequest",
            "ReturnReqCreationRequest"
        ),
    toAdminDecisionRequest: (model: ReturnRequest): AdminDecisionRequest =>
        returnRequestMapper.map<ReturnRequest, AdminDecisionRequest>(
            model,
            "ReturnRequest",
            "AdminDecisionRequest"
        ),
    toLesseeSubmitRequest: (model: ReturnRequest): LesseeSubmitReturnRequest =>
        returnRequestMapper.map<ReturnRequest, LesseeSubmitReturnRequest>(
            model,
            "ReturnRequest",
            "LesseeSubmitReturnRequest"
        ),
    toLessorConfirmRequest: (model: ReturnRequest): LessorConfirmReturnRequest =>
        returnRequestMapper.map<ReturnRequest, LessorConfirmReturnRequest>(
            model,
            "ReturnRequest",
            "LessorConfirmReturnRequest"
        ),
};
